rllib/core/models/base.py

```diff
@@ -0,0 +1,380 @@
+
+        Returns:
+            NestedDict: The output tensors.
+        """
+        raise NotImplementedError
+
+
+class ActorCriticEncoder(Encoder):
+    """An encoder that potentially holds two encoders.
+
+    This is a special case of encoder that can either enclose a single,
+    shared encoder or two separate encoders: One for the actor and one for the
+    critic. The two encoders are of the same type and we can therefore make the
+    assumption that they have the same input and output specs.
+    """
+
+    framework = None
+
+    def __init__(self, config: ModelConfig) -> None:
+        if config.shared:
+            self.encoder = config.base_encoder_config.build(framework=self.framework)
+        else:
+            self.actor_encoder = config.base_encoder_config.build(
+                framework=self.framework
+            )
+            self.critic_encoder = config.base_encoder_config.build(
+                framework=self.framework
+            )
+
+        # We need to call Encoder.__init__() after initializing the encoder(s) in
+        # order to build on their specs.
+        super().__init__(config)
+
+    @override(Model)
+    def get_input_spec(self) -> Union[Spec, None]:
+        if self.config.shared:
+            state_in_spec = self.encoder.input_spec[STATE_IN]
+        else:
+            state_in_spec = NestedDict(
+                {
+                    ACTOR: self.actor_encoder.input_spec[STATE_IN],
+                    CRITIC: self.critic_encoder.input_spec[STATE_IN],
+                }
+            )
+
+        return SpecDict(
+            {
+                SampleBatch.OBS: None,
+                STATE_IN: state_in_spec,
+                SampleBatch.SEQ_LENS: None,
+            }
+        )
+
+    @override(Model)
+    def get_output_spec(self) -> Union[Spec, None]:
+        if self.config.shared:
+            state_out_spec = self.encoder.output_spec[STATE_OUT]
+        else:
+            state_out_spec = NestedDict(
+                {
+                    ACTOR: self.actor_encoder.output_spec[STATE_OUT],
+                    CRITIC: self.critic_encoder.output_spec[STATE_OUT],
+                }
+            )
+
+        return SpecDict(
+            {
+                ENCODER_OUT: {
+                    ACTOR: None,
+                    CRITIC: None,
+                },
+                STATE_OUT: state_out_spec,
+            }
+        )
+
+    @override(Model)
+    def get_initial_state(self):
+        if self.config.shared:
+            return self.encoder.get_initial_state()
+        else:
+            return {
+                ACTOR: self.actor_encoder.get_initial_state(),
+                CRITIC: self.critic_encoder.get_initial_state(),
+            }
+
+    @override(Model)
+    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+        if self.config.shared:
+            outs = self.encoder(inputs, **kwargs)
+            return NestedDict(
```

Comment:
Just FYI. You can return dicts and they will get converted to NestedDict due to spec checking decoration

Reply:
Thanks! Yeah, I just did this because it would otherwise violate the return type.